File: syncflow-java/syncflow-workflow/src/main/java/com/syncflow/workflow/service/WorkflowService.java
package com.syncflow.workflow.service;

import com.syncflow.workflow.dto.ApprovalCommentVO;
import com.syncflow.workflow.dto.ApprovalTaskVO;
import com.syncflow.workflow.dto.BusinessObjectVO;

import java.util.List;

/**
 * Core workflow service interface.
 * <p>
 * Provides a generic approval API that binds ANY business entity to a Flowable
 * process. Callers only need to supply a process key and the target object
 * metadata; the workflow engine handles routing, task assignment, and history.
 */
public interface WorkflowService {

    /**
     * Start an approval process for a business object.
     *
     * @param processKey  Flowable process-definition key
     * @param objectId    primary key of the business entity
     * @param objectType  business type, e.g. TASK, BOM
     * @param objectName  human-readable display name
     * @param projectId   owning project id
     * @param applicantId the user initiating the approval
     * @return the id of the newly created {@code wf_business_object} record
     */
    Long startProcess(String processKey, Long objectId, String objectType,
                      String objectName, Long projectId, Long applicantId);

    /**
     * Start an approval process with optional CC users.
     *
     * @param ccUserIds user ids to CC on this approval (may be null or empty)
     */
    Long startProcess(String processKey, Long objectId, String objectType,
                      String objectName, Long projectId, Long applicantId,
                      java.util.List<Long> ccUserIds);

    /**
     * Complete an approval task (approve or reject).
     *
     * @param taskId     Flowable user-task id
     * @param approverId the user completing the task
     * @param approved   {@code true} to approve, {@code false} to reject
     * @param comment    optional comment
     */
    void completeTask(String taskId, Long approverId, boolean approved, String comment);

    /**
     * Get pending approval tasks assigned to a user.
     *
     * @param userId target user id
     * @return list of pending tasks
     */
    List<ApprovalTaskVO> getPendingTasks(Long userId);

    /**
     * Get completed/processed approval tasks for a user (tasks they have acted on).
     *
     * @param userId the user id
     * @return list of completed approval tasks
     */
    List<ApprovalTaskVO> getCompletedTasks(Long userId);

    /**
     * Get the full approval history for a business object.
     *
     * @param businessObjectId FK to wf_business_object.id
     * @return ordered list of approval comments / actions
     */
    List<ApprovalCommentVO> getApprovalHistory(Long businessObjectId);

    /**
     * Get the business object detail.
     *
     * @param id FK to wf_business_object.id
     * @return business object VO, or {@code null} if not found
     */
    BusinessObjectVO getBusinessObject(Long id);

    /**
     * Get the business object entity (for internal cross-module use).
     */
    com.syncflow.workflow.entity.BusinessObject getBusinessObjectEntity(Long id);

    /**
     * Find a business object by object type and object ID.
     */
    com.syncflow.workflow.entity.BusinessObject findBusinessObject(String objectType, Long objectId);

    /**
     * Withdraw a pending approval.
     *
     * @param businessObjectId FK to wf_business_object.id
     * @param userId           the user requesting the withdrawal (must be the applicant)
     */
    void withdrawApproval(Long businessObjectId, Long userId);

    /**
     * Withdraw a pending approval by its Flowable process instance ID.
     * Used when a task is manually completed while an approval is still pending.
     *
     * @param flowInstanceId the Flowable process instance ID
     */
    void withdrawByFlowInstanceId(String flowInstanceId);

    /**
     * Send a reminder notification to the current assignee of a pending approval.
     *
     * @param businessObjectId the business object to remind about
     * @param userId           the user requesting the reminder (must be the applicant)
     */
    void remindApproval(Long businessObjectId, Long userId);

    /**
     * Add a candidate user (加签) to an active Flowable task.
     *
     * @param taskId          the Flowable user-task id
     * @param candidateUserId the user to add as a candidate
     */
    void addCandidateUser(String taskId, Long candidateUserId);

    /**
     * Check if approval is enabled for a given object type.
     *
     * @param objectType the business object type (e.g. PROJECT, TASK)
     * @return true if approval is configured and enabled for this type
     */
    boolean isApprovalRequired(String objectType);

    /**
     * Reassign (transfer) an active Flowable task to a different user.
     *
     * @param taskId    the Flowable user-task id
     * @param newUserId the user to reassign to
     */
    void reassignTask(String taskId, Long newUserId);
}

File: src/services/workflow.service.ts
import api from './api';

export interface ApprovalTaskVO {
  taskId: string;
  taskName: string;
  businessObjectId: number;
  objectType: string;
  objectName: string;
  objectCode?: string;
  projectId?: number;
  applicantName?: string;
  createdAt: string;
}

export interface ApprovalCommentVO {
  id: number;
  nodeName: string;
  approverName: string;
  action: string; // APPROVE/REJECT
  comment?: string;
  createdAt: string;
}

export interface BusinessObjectVO {
  id: number;
  objectType: string;
  objectId: number;
  objectName: string;
  objectCode?: string;
  projectId?: number;
  status: number; // 1=draft, 2=pending, 3=approved, 4=rejected, 5=withdrawn
  currentNode?: string;
  currentTaskId?: string;
  flowInstanceId?: string;
  applicantId: number;
  applicantName?: string;
  appliedAt?: string;
  completedAt?: string;
}

export interface DelegationVO {
  id: number;
  businessObjectId: number;
  fromUserId: number;
  toUserId: number;
  reason?: string;
  startTime?: string;
  endTime?: string;
  isActive: boolean;
  createdAt: string;
}

export interface CcRecordVO {
  id: number;
  businessObjectId: number;
  userId: number;
  isRead: boolean;
  readAt?: string;
  createdAt: string;
}

// Get business object detail by id
export async function getBusinessObject(id: string | number): Promise<BusinessObjectVO> {
  return api.get(`/wf/business-objects/${id}`);
}

// Complete approval task (approve or reject)
export async function completeTask(taskId: string, data: {
  approved: boolean;
  comment?: string;
}) {
  return api.post(`/wf/tasks/${taskId}/complete`, data);
}

// Start a workflow instance for a business object
export async function startWorkflow(data: {
  processKey: string;
  objectId: number;
  objectType: string;
  objectName: string;
  projectId?: number;
  ccUserIds?: number[];
}) {
  return api.post('/wf/start', data);
}

// Get pending approval tasks
export async function getPendingTasks(): Promise<ApprovalTaskVO[]> {
  return api.get('/wf/tasks/pending');
}

// Get completed/processed approval tasks
export async function getCompletedTasks(): Promise<ApprovalTaskVO[]> {
  return api.get('/wf/tasks/completed');
}

// Get approval history
export async function getApprovalHistory(businessObjectId: number): Promise<ApprovalCommentVO[]> {
  return api.get(`/wf/business-objects/${businessObjectId}/history`);
}

// Withdraw approval
export async function withdrawApproval(businessObjectId: number) {
  return api.post(`/wf/business-objects/${businessObjectId}/withdraw`);
}

// Send reminder to current assignee
export async function remindApproval(businessObjectId: number) {
  return api.post(`/wf/business-objects/${businessObjectId}/remind`);
}

// Delegation
export async function delegate(data: {
  businessObjectId: number;
  fromUserId: number;
  toUserId: number;
  reason?: string;
  startTime?: string;
  endTime?: string;
}) {
  return api.post('/wf/delegation', data);
}

export async function revokeDelegation(id: number) {
  return api.delete(`/wf/delegation/${id}`);
}

export async function getDelegations(): Promise<DelegationVO[]> {
  return api.get('/wf/delegation');
}

// CC Records
export async function addCc(data: { businessObjectId: number; userId: number }) {
  return api.post('/wf/cc', data);
}

// Add candidate user (加签) to an active task
export async function addCandidateUser(taskId: string, userId: number) {
  return api.post(`/wf/tasks/${taskId}/add-candidate`, { userId });
}

// Reassign (transfer) an active task to a different user
export async function reassignTask(taskId: string, userId: number) {
  return api.post(`/wf/tasks/${taskId}/reassign`, { userId });
}

export async function markCcAsRead(id: number) {
  return api.put(`/wf/cc/${id}/read`);
}

export async function getCcRecords(unreadOnly = false): Promise<CcRecordVO[]> {
  return api.get(`/wf/cc?unreadOnly=${unreadOnly}`);
}
